frontend/src/metabase/visualizations/visualizations/Table.jsx

```diff
@@ -175,29 +175,44 @@ export default class Table extends Component {
       title: t`Columns`,
       widget: ChartSettingOrderedSimple,
       getHidden: (_series, vizSettings) => vizSettings["table.pivot"],
-      isValid: ([{ card, data }], _vizSettings, extra = {}) =>
+      getValue: ([{ card, data }], vizSettings, extra = {}) => {
         // If "table.columns" happened to be an empty array,
         // it will be treated as "all columns are hidden",
         // This check ensures it's not empty,
         // otherwise it will be overwritten by `getDefault` below
-        card.visualization_settings["table.columns"].length !== 0 &&
-        (extra.isQueryRunning ||
-          _.all(
-            card.visualization_settings["table.columns"],
-            columnSetting =>
-              !columnSettings.enabled ||
-              findColumnIndexForColumnSetting(data.cols, columnSetting) >= 0,
-          )),
-      getDefault: ([
-        {
-          data: { cols },
-        },
-      ]) =>
-        cols.map(col => ({
-          name: col.name,
-          fieldRef: col.field_ref,
-          enabled: col.visibility_type !== "details-only",
-        })),
+        const tableColumns = vizSettings["table.columns"];
+
+        const isValid =
+          tableColumns &&
+          tableColumns.length !== 0 &&
+          (extra.isQueryRunning ||
+            _.all(
+              tableColumns,
+              columnSetting =>
+                !columnSettings.enabled ||
+                findColumnIndexForColumnSetting(data.cols, columnSetting) >= 0,
+            ));
+        if (!isValid) {
+          return data.cols.map(col => ({
+            name: col.name,
+            fieldRef: col.field_ref,
+            enabled: col.visibility_type !== "details-only",
+          }));
+        }
+
+        // Remove any columnSettings that do not appear in data.cols
+        const filteredTableColumns = tableColumns.filter(columnSetting =>
```

Comment:
nit 

We might want to call this const something like `tableColumnsPresentInDataCols`…or something kind of like that.

Right now the declaration and the use of the const are nearby, in the future these may spread apart and the name could last longer than that :))